Migrate CI to bionic

mikaelarguedas opened the pull request:
start migrating Linux CI to bionic

connects to ros2/ros2#481

dirk-thomas commented:
I would suggest to split the bullets into two groups:
* what needs to happen to get this merged and use Bionic in the CI jobs
* what should happen afterwards

mikaelarguedas commented:
> I would suggest to split the bullets into two groups:

> what needs to happen to get this merged and use Bionic in the CI jobs
> what should happen afterwards

Agreed, though it depends what we consider "enough" to merge this on CI.
If we want all our nightly to build on bionic without warnings we will need almost all of them to be addressed.

I see several stages:
- Being able to run our regular ci_linux
- Being able to run all the nightlies (including tb2 demo jobs)
- Being able to run all ci jobs including the ones using opensplice.
- Audit dependencies and previous hacks

I think we need at least the first 2 to be done before merging this on CI. Do you agree?

dirk-thomas commented:
> I think we need at least the first 2 to be done before merging this on CI. Do you agree?

Yes, except that I don't see the tb2 demo as a blocker - these jobs could continue to use Xenial until `kobuki` has been released.

nuclearsandwich commented:
Are we going to drop support for xenial on HEAD the moment that bionic is working or will there be a period of time where we need to support both with the same branch? Of course, ardent will remain a release supported on xenial until bouncy is out. I'm just wondering if we need to test both configurations at once or if we expect all ROS 2 developers to jump to bionic instantly.

mikaelarguedas commented:
Status update:
This branch with connected PRs gives green CI for both opensplice and connext including cross_vendor tests between the 2.

Fast-RTPS still segfaults on any operation involving setting parameters or security

mikaelarguedas commented:
New round of packaging job including the warning fixups from https://github.com/ros2/system_tests/pull/269 and https://github.com/ros2/rosidl/pull/275

* Packaging amd64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=test_packaging_bionic_linux&build=1)](http://ci.ros2.org/job/test_packaging_bionic_linux/1/)
* Packaging Windows: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_packaging_windows&build=40)](https://ci.ros2.org/job/ci_packaging_windows/40/)
* Packaging MacOS: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_packaging_osx&build=33)](https://ci.ros2.org/job/ci_packaging_osx/33/)

mikaelarguedas commented:
Squashed and rebased all commits.

Ready for review

Last round of CI to confirm all is well:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4369)](http://ci.ros2.org/job/ci_linux/4369/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=test_bionic_aarch64&build=2)](http://ci.ros2.org/job/test_bionic_aarch64/2/) (custom job)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3572)](http://ci.ros2.org/job/ci_osx/3572/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4428)](http://ci.ros2.org/job/ci_windows/4428/)

* Linux Debug [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4369)](http://ci.ros2.org/job/ci_linux/4370/)
* Linux Release [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4369)](http://ci.ros2.org/job/ci_linux/4371/)

* Coverage: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=test_bionic_cov&build=2)](http://ci.ros2.org/job/test_bionic_cov/2/) (custom job)

* Packaging aarch64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=test_bionic_aarch64_packaging&build=2)](http://ci.ros2.org/job/test_bionic_aarch64_packaging/2/) (custom job)
* Packaging amd64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=test_packaging_bionic_linux&build=2)](http://ci.ros2.org/job/test_packaging_bionic_linux/2/) (custom job)

mikaelarguedas commented:
Looking a bit more into it we were **not** using custom arm images but the images from that project: https://hub.docker.com/r/aarch64/ubuntu/

Now that the official docker library/ubuntu hosts armhf and arm64 images I kept my change as is to use the upstream docker images.
If the need to provide CI that can run with qemu comes up we move to use our custom images. In the meantime we'll keep upstream to avoid having custom "magic"

nuclearsandwich commented:
> Now that the official docker library/ubuntu hosts armhf and arm64 images I kept my change as is to use the upstream docker images.
> If the need to provide CI that can run with qemu comes up we move to use our custom images. In the meantime we'll keep upstream to avoid having custom "magic"

Makes complete sense to me. If there's no qemu support to preserve then there's no reason not to go straight to upstream images.

nuclearsandwich commented on linux_docker_resources/Dockerfile:
```diff
@@ -69,13 +71,14 @@ RUN apt-get update && apt-get install --no-install-recommends -y libasio-dev lib
-# We are building poco from source for now as we need at least 1.4.1p1 and xenial ships with 1.3.6p1 (https://github.com/ros2/poco_vendor/pull/10)
-# RUN apt-get update && apt-get install --no-install-recommends -y libpoco-dev libpocofoundation9v5 libpocofoundation9v5-dbg
+# Install build dependencies for class_loader.
+# TODO(mikaelarguedas) Uncomment this when we decide what to do about Poco debug builds
```
Are we cool to merge with this TODO still pending?

mikaelarguedas replied:
TL;DR: IMO yes, this doesnt change the current behavior ATM.

Poco doesn't ship debs with debug libraries anymore (Bionic or Brew) and the release libraries will be found even in Debug builds.
We should be able to just use system [poco on Ubuntu] (https://ci.ros2.org/job/ci_packaging_linux/91/) and MacOS and build in Debug on top without facing https://github.com/ros2/poco_vendor/issues/11.
@sloretz is currently modifying his [poco_vendor PR](https://github.com/ros2/poco_vendor/pull/14) to build only in Release and not overriding the Poco CMake module, my hope is that this should work on windows and allow us to build poco, only on Windows and only in Release that should save use some CI time.

mikaelarguedas commented:
Status update:
  - ros2.repos:
    - builds successfully with all rmw implementations for all job types on both arm64 and amd64
    - outstanding issues:
      - currently rviz doesnt build on bionic if system yaml-cpp is installed: https://github.com/ros2/rviz/pull/160
  - turtlebot2_demo.repos:  [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_bionic_turtlebot-demo_linux&build=13)](https://ci.ros2.org/job/ci_bionic_turtlebot-demo_linux/13/)
    - Outstanding issues:
      - astra camera: some warnings that need a patch stage in CMake to be silenced when building OpenNi, ticketed upstream: https://github.com/orbbec/OpenNI2/issues/5
      - map_server: some tests fail to compile: 

    - Rebasing status:
      - Deprecating our repo:
        - cartographer: use packaged 0.3.0 in melodic, deprecate our fork
      - Rebased on top of upstream:
        - cartographer_ros: rebased on top of 0.3.0
        - depthimage_to_laserscan: already up to date
        - joy: rebased on top of latest indigo-devel
        - teleop_twist_joy: up to date
      - Not rebased:
        - navigation: not rebased
        - pcl_conversion: not rebased as upstream repo now empty
        - ros_astra_camera: not rebased
        - teleop_twist_keyboard: not rebased
        - vision_opencv: not rebased

dhood commented on linux_docker_resources/Dockerfile:
```diff
@@ -89,7 +89,18 @@ RUN echo "@today_str"
+RUN if test ${BRIDGE} = true; then apt-get update && apt-get install --no-install-recommends -y \
+    python-rospkg \
+    python3-crypto \
+    python3-gnupg \
```
I'm curious why we bring in the dependencies manually here instead of installing the ros package that needs them?

mikaelarguedas replied:
the ros package installs the python 2 version of these dependencies but for some reason I needed the python3 version for this to work in my container.

Currently, while installing the python3 version works locally, neither work on CI so I need to find another solution for the `Crypto` module to be found properly

dhood commented:
Packaging job on bionic is pending decision in https://github.com/ros2/ros1_bridge/issues/118

mikaelarguedas commented:
New round of CI using the last state of this branch and the newly generated jobs with the ROS1DISTRo selector:

* Linux None: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=mikael_test_ci_linux&build=1)](https://ci.ros2.org/view/All/job/mikael_test_ci_linux/1/)
* Linux Debug: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=mikael_test_ci_linux&build=2)](https://ci.ros2.org/view/All/job/mikael_test_ci_linux/2/)
* Linux Release (using https://github.com/ros2/rosidl/pull/275): [![Build Status](http://ci.ros2.org/buildStatus/icon?job=mikael_test_ci_linux&build=6)](https://ci.ros2.org/view/All/job/mikael_test_ci_linux/6/)
* Linux None (with opensplice): [![Build Status](http://ci.ros2.org/buildStatus/icon?job=mikael_test_ci_linux&build=4)](https://ci.ros2.org/view/All/job/mikael_test_ci_linux/4/)

* Linux aarch64 None: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=mikael_test_ci_linux-aarch64&build=1)](https://ci.ros2.org/view/All/job/mikael_test_ci_linux-aarch64/1/)
* Linux aarch64 Debug: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=mikael_test_ci_linux-aarch64&build=2)](https://ci.ros2.org/view/All/job/mikael_test_ci_linux-aarch64/2/)
* Linux aarch64 Release (using https://github.com/ros2/rosidl/pull/275): [![Build Status](http://ci.ros2.org/buildStatus/icon?job=mikael_test_ci_linux-aarch64&build=5)](https://ci.ros2.org/view/All/job/mikael_test_ci_linux-aarch64/5/)
* Linux aarch64 None (with opensplice): [![Build Status](http://ci.ros2.org/buildStatus/icon?job=mikael_test_ci_linux-aarch64&build=4)](https://ci.ros2.org/view/All/job/mikael_test_ci_linux-aarch64/4/)

* Coverage: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=mikael_test_linux_coverage&build=1)](http://ci.ros2.org/job/mikael_test_linux_coverage/1/) (custom job)

* Packaging amd64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=mikael_test_ci_packaging_linux&build=2)](http://ci.ros2.org/job/mikael_test_ci_packaging_linux/2/) 
* Packaging aarch64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=mikael_test_ci_packaging_linux-aarch64&build=3)](http://ci.ros2.org/job/mikael_test_ci_packaging_linux-aarch64/3/)

mikaelarguedas commented:
> not sure if you were looking for "fresh" eyes on it though

Not necessarily :) as we worked on different parts of the PR I think that's fine :+1: 

It feels a bit weird to have the `CI_UBUNTU_DISTRO` option now displayed on all jobs  including the Macos/Windows ones but I think that's the price to pay for having a single template for all of them.

I'm planning on merging this as is in ~1hr.

If you happen to want to review and this is already merged, please feel free to comment here and I'll address comments post merge in a follow-up.

mikaelarguedas commented:
And some jobs for the turtlebot demo:
* tb2 amd64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=mikael_ci_turtlebot-demo_linux&build=5)](http://ci.ros2.org/job/mikael_ci_turtlebot-demo_linux/5/)
* tb2 aarch64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=mikael_ci_turtlebot-demo_linux-aarch64&build=6)](http://ci.ros2.org/job/mikael_ci_turtlebot-demo_linux-aarch64/6/)

mikaelarguedas commented:
Nightly turtlebot jobs failed due to an [outstanding PR](https://github.com/ros2/ros_astra_camera/pull/28) that wasn't merged. This is now fixed

* tb2 amd64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_turtlebot-demo_linux&build=131)](http://ci.ros2.org/job/ci_turtlebot-demo_linux/131/)
* tb2 aarch64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_turtlebot-demo_linux-aarch64&build=112)](http://ci.ros2.org/job/mikael_ci_turtlebot-demo_linux-aarch64/112/)
